
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://RamonZwaan.github.io/metapepview/experiment-evaluation/">
      
      
        <link rel="prev" href="../experiment-performances-biases/">
      
      
        <link rel="next" href="../build-reference-dataset/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.17">
    
    
      
        <title>Spectral evaluation - MetaPepView Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.7e37652d.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#spectral-data-analysis-and-experiment-benchmarking" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="MetaPepView Documentation" class="md-header__button md-logo" aria-label="MetaPepView Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            MetaPepView Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Spectral evaluation
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="MetaPepView Documentation" class="md-nav__button md-logo" aria-label="MetaPepView Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    MetaPepView Documentation
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Introduction
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../getting-started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting started
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../prepare-input-data/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Prepare input data
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    User guide
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            User guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" >
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Data management
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            Data management
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../data-import/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Data import
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../annotation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Annotation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../project-table/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Project table
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2" >
        
          
          <label class="md-nav__link" for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Community visualization
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            Community visualization
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../community-visualization/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../taxonomy-visualization/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Community composition
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../function-visualization/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Community function
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" checked>
        
          
          <label class="md-nav__link" for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Experiment evaluation
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            Experiment evaluation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../experiment-performances-biases/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Spectral evaluation
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Spectral evaluation
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#data-import" class="md-nav__link">
    <span class="md-ellipsis">
      Data import
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#benchmarking" class="md-nav__link">
    <span class="md-ellipsis">
      Benchmarking
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Benchmarking">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#confidence-thresholds-barplot" class="md-nav__link">
    <span class="md-ellipsis">
      Confidence thresholds barplot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#peptide-confidence-scatterplot" class="md-nav__link">
    <span class="md-ellipsis">
      Peptide confidence scatterplot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#confidence-ranked-distribution" class="md-nav__link">
    <span class="md-ellipsis">
      Confidence ranked distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scan-intensity-distribution" class="md-nav__link">
    <span class="md-ellipsis">
      Scan intensity distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ion-transmission" class="md-nav__link">
    <span class="md-ellipsis">
      Ion Transmission
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#miscleavage-distribution" class="md-nav__link">
    <span class="md-ellipsis">
      Miscleavage distribution
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#experimental-quality" class="md-nav__link">
    <span class="md-ellipsis">
      Experimental quality
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Experimental quality">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#general-experiment-metrics" class="md-nav__link">
    <span class="md-ellipsis">
      General experiment metrics
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#total-ion-current" class="md-nav__link">
    <span class="md-ellipsis">
      Total Ion Current
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mz-over-rt" class="md-nav__link">
    <span class="md-ellipsis">
      m/z over RT
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ms2-ion-transmission" class="md-nav__link">
    <span class="md-ellipsis">
      MS2 Ion Transmission
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scan-intensities" class="md-nav__link">
    <span class="md-ellipsis">
      Scan Intensities
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#charge-distribution" class="md-nav__link">
    <span class="md-ellipsis">
      Charge distribution
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../build-reference-dataset/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Build reference dataset
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../evaluate-taxonomy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Evaluate community composition
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    MetaPepView backend
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            MetaPepView backend
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../annotation-pipeline/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Annotation pipeline
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#data-import" class="md-nav__link">
    <span class="md-ellipsis">
      Data import
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#benchmarking" class="md-nav__link">
    <span class="md-ellipsis">
      Benchmarking
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Benchmarking">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#confidence-thresholds-barplot" class="md-nav__link">
    <span class="md-ellipsis">
      Confidence thresholds barplot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#peptide-confidence-scatterplot" class="md-nav__link">
    <span class="md-ellipsis">
      Peptide confidence scatterplot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#confidence-ranked-distribution" class="md-nav__link">
    <span class="md-ellipsis">
      Confidence ranked distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scan-intensity-distribution" class="md-nav__link">
    <span class="md-ellipsis">
      Scan intensity distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ion-transmission" class="md-nav__link">
    <span class="md-ellipsis">
      Ion Transmission
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#miscleavage-distribution" class="md-nav__link">
    <span class="md-ellipsis">
      Miscleavage distribution
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#experimental-quality" class="md-nav__link">
    <span class="md-ellipsis">
      Experimental quality
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Experimental quality">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#general-experiment-metrics" class="md-nav__link">
    <span class="md-ellipsis">
      General experiment metrics
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#total-ion-current" class="md-nav__link">
    <span class="md-ellipsis">
      Total Ion Current
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mz-over-rt" class="md-nav__link">
    <span class="md-ellipsis">
      m/z over RT
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ms2-ion-transmission" class="md-nav__link">
    <span class="md-ellipsis">
      MS2 Ion Transmission
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scan-intensities" class="md-nav__link">
    <span class="md-ellipsis">
      Scan Intensities
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#charge-distribution" class="md-nav__link">
    <span class="md-ellipsis">
      Charge distribution
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="spectral-data-analysis-and-experiment-benchmarking">Spectral data analysis and experiment benchmarking</h1>
<p>MetaPepView provides a separate module for visualization of (meta)proteomics experimental performance parameters, including benchmarking of experiments against a dataset of metaproteomics experiments. This visualization toolbox is present in the <em>Experiment performance evaluation</em> module. Here, MS spectral data can be imported in <a href="../prepare-input-data/#spectral-data-mzml"><em>mzML</em> format</a>, additional <em>feature</em> information may be added in <a href="../prepare-input-data/#feature-data-featurexml"><em>featureXML</em> format</a>, and spectral data may be supplemented with DB search and <em>de novo</em> (meta)proteomics data (<a href="../data-import/#import-components">several formats supported</a>). This allows visualization of spectral quality parameters, as well as to link these parameters to peptide identification rates.</p>
<p><img alt="QA_module.PNG" src="../images/QA_module.PNG" /></p>
<p>The Quality control module is structured in several tabs. The <em>Benchmarking</em> tab provides visualizations to compare (benchmark) an imported experiment with a experimental benchmark dataset, the <em>Experimental quality</em> tab provides visualizations connecting spectral quality to peptide identification performance.</p>
<h2 id="data-import">Data import</h2>
<p>On the top of the module page, importers for the spectral data (<em>mzML</em>), feature data (<em>featureXML</em>), DB search data and <em>de novo</em> data are provided. Data imported here will be visualized in the module. Part of the datasets may be omitted on import. However, some visualizations that require the missing datasets will not be displayed. For DB search and <em>de novo</em> data, it should be ensured that they were derived from the same MS run as the <em>mzML</em> and <em>featureXML</em> file. MetaPepView will only combine data that are from the same MS run.</p>
<h2 id="benchmarking">Benchmarking</h2>
<p>The benchmarking tab provides a set of visualizations that compare a single experiment against a dataset of experiments. MetaPepView provides two standard datasets, shown under the <em>Reference dataset</em> dropdown menu, for testing purposes. However, it is good practice to <a href="../build-reference-dataset/">generate your own benchmark datasets</a> with suitable samples and MS conditions. </p>
<p>A new benchmark dataset can be imported from the <em>import custom dataset</em> importer. A CLI tool "mpv-buildref" is provided to build a new benchmark dataset. When providing a root directory (which should contain all <em>mzML</em> files, together with DB search and <em>de novo</em> data), as well as format settings, this tool will generate a summary file that contains performance parameters for all experiments that can be imported into the module.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>To prevent biases in proteomics data processing between samples, MetaPepView  will only benchmark experiments against the reference dataset if the DB search and <em>de novo</em> input data are in the same format for both dataset. The formats for the benchmark dataset is displayed after import into the dashboard.</p>
<p>This format constraint is also present for the benchmark dataset. If a custom dataset is constructed, all DB search and <em>de novo</em> files should be in the same format across all experiments.</p>
</div>
<p><img alt="QA_benchmark_module.PNG" src="../images/QA_benchmark_module.png" /></p>
<p>An overview of the available evaluation graphs is shown below.</p>
<h3 id="confidence-thresholds-barplot">Confidence thresholds barplot</h3>
<p>Quick profiling of the output quality from a (meta)proteomics experiment can be performed by comparing the number (or fraction) of peptide identifications above predefined confidence thresholds. The thresholds from a single experiment (left bar) can be compared to the experiments from the benchmark dataset (right bars). The graph allows scaling of the y-axis by peptide counts, fraction matches above thresholds of total scans, and the average number of matches above thresholds per minute (scaled by total retention time). For <em>de novo</em> identifications (bottom bars), either all identifications may be shown, or they may be filtered to only show <em>de novo</em> peptides not identified from DB search.</p>
<p>While the graph allows quick observation of identification performance between experiments, certain patterns may imply issues in the experimental/analysis conditions:</p>
<ul>
<li>A low fraction of high confidence peptide identifications compared to lower confidences may imply lower spectral quality from the raw MS data. This should also be reflected in a low number of <em>de novo</em> identifications.</li>
<li>A low number of DB search matches combined with a large number of (high confidence) <em>de novo</em> identifications may be due to the use of a database for DB search that does not represent the metaproteome of the sample well. Also, presence of high confidence <em>de novo only</em> peptides (peptides that were not identified from DB search matching) implies that.</li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The numbers/fractions of DB search matches and <em>de novo</em> identifications that may be considered acceptable vary greatly between the type of samples and analysis conditions. Therefore, it is important that the benchmark dataset comprise experiments similar to the analysed sample.</p>
</div>
<p><img alt="QA_bench_bars.jpg" src="../images/QA_bench_bars.jpg" /></p>
<h3 id="peptide-confidence-scatterplot">Peptide confidence scatterplot</h3>
<p>An alternative method to observe the number of peptide identifications of an experiment (red markers) against the benchmark dataset (blue dots) is through a scatterplot. This graph shows the number (or fraction) of peptide identifications above predefined thresholds from DB search matching (left groups), <em>de novo</em> identifications (middle groups), and <em>de novo</em> only (d-only) identifications (right groups). Peptide identifications may be normalized by fraction of total MS2 scans, or by total retention time (count per minute).</p>
<p><img alt="QA_bench_scatter.PNG" src="../images/QA_bench_scatter.PNG" /></p>
<h3 id="confidence-ranked-distribution">Confidence ranked distribution</h3>
<p>The distribution of peptide identification confidence between an experiment (red line) and the benchmark dataset (gray area for 95% confidence interval) can be observed in the ranked distribution plot. Here, peptide identifications are sorted on the x-axis by confidence (y-axis). The distribution (x-axis) can be shown as top-n matches, or normalized as fraction of total matches, or fraction of MS2 matches. The distribution of either DB search or <em>de novo</em> identifications may be plotted.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>DB search and <em>de novo</em> tools often only report peptides above a confidence threshold, unless <em>x-axis normalization</em> is set to <em>Total matches</em>, the distribution of the benchmark dataset will be skewed as more experiments reach the confidence threshold (the end of the dataset). This should be taken into account when analysing the further end of the x-axis.</p>
</div>
<p><img alt="conf_rank_dist.png" src="../images/conf_rank_dist.png" /></p>
<h3 id="scan-intensity-distribution">Scan intensity distribution</h3>
<p>One metric for spectral quality evaluation is the distribution of MS scan intensities. Here, this is plotted as the median, 90'th percentile, and 99'th percentile intensity of MS1 and MS2 scan. An experimental sample (red marker) may be compared against the benchmark dataset (blue dots). </p>
<p>Several factors may influence the intensity distribution of scans. For example, low MS1 intensities may imply that low amount of sample was submitted to the MS. Low MS2 intensity (relative to MS1) may imply low ion transmission efficiency (large loss of signal during fragmentation). Finally, strange patterns between median and <em>x</em> percentile groups may indicate presence of pollutants, or sparse elution of peptides. </p>
<p><img alt="QA_benbch_ints.PNG" src="../images/QA_benbch_ints.PNG" /></p>
<h3 id="ion-transmission">Ion Transmission</h3>
<p>Presence of pollutants during MS analysis or insufficient cleaning of the MS may result in greater loss of signal during ion transmission, especially during fragmentation of peptides prior to MS2 analysis. This may greatly impact the quality of (meta)proteomics data as MS2 signal loss lowers the confidence of peptide identification.</p>
<p>Here, the distribution of ion transmission efficiencies of MS2 scans is visualized as total MS2 signal as percentage of the precursor ion measured in the prior MS1 scan (intensity of the peak from MS1, which was selected for MS2 scanning):</p>
<div class="arithmatex">\[
Eff_{ion\ tr} = \frac{Int_{tic\ MS2}}{Int_{prec\ MS1}} \times 100\%
\]</div>
<p>The median efficiency, as well as the 90'th and 99'th percentile (lowest efficiencies) are visualized. A high percentage indicates higher transmission efficiency. Thus, a larger fraction of total signal was measured in MS2.</p>
<p>The transmission efficiency may be <em>rescaled by ion injection time</em>; a major factor of signal intensity is the accumulation time of ions prior to the MS scan, where shorter injection times result in lower signal intensities. This time may change on a per-scan basis to prevent oversaturation of the sensors. Generally, the injection time of MS1 scans are much shorter than those from MS2 scans (which may result in transmission efficiencies &gt;100%). In addition, MS1 injection times may vary strongly. By applying <em>Scale ion injection time</em>, all scans are corrected for their injection times before the ion transmission is calculated: 
$$
Eff_{ion\ tr} = \frac{Int_{tic\ MS2}}{Int_{prec\ MS1}} \times \frac{Inj_{MS1}}{Inj_{MS2}} \times 100\%
$$  </p>
<p><img alt="QA_bench_Ion_Transm.PNG" src="../images/QA_bench_Ion_Transm.PNG" /></p>
<h3 id="miscleavage-distribution">Miscleavage distribution</h3>
<p>Prior to MS submission, proteins are cleaved into short peptides, often with Trypsin (in combination with Lys-C) enzyme. These cleave proteins into peptides that have lengths well suited for MS detection and identification. This protein digestion is usually a time consuming step, and often, a balance is made between the amount of cleavage and digestion time. However, too short digestion time, as well as potential bad conditions during digestion, may result in a large fraction of multiply miscleaved peptides. As a result, a large fraction of protein material may become inaccessible for MS detection.</p>
<p>This graph analyses DB search data from the reference dataset and the experiment dataset to provide a distribution of miscleavages, it counts the number of cleavage amino acids (arginine/lysine) present in each peptide (Except for the C-terminal amino-acid) and counts PSM's at each miscleavage number. These are provided as stacked bar graphs.</p>
<p><img alt="miscleave_dist.PNG" src="../images/miscleave_dist.png" /></p>
<h2 id="experimental-quality">Experimental quality</h2>
<p>The experimental quality page shows detailed information of spectral and (meta)proteomics performance for a single sample.</p>
<p><img alt="QA_exp_perf_module.PNG" src="../images/QA_exp_perf_module.png" /></p>
<h3 id="general-experiment-metrics">General experiment metrics</h3>
<p>Key metrics from a single MS experiment are displayed at the top of the page. These metrics give a quick overview of the experiment. These include:</p>
<ul>
<li><em>Total retention time (min)</em>: Total length of the MS run.</li>
<li><em>MS1 scans</em>: The number of MS1 (precursor) scans.</li>
<li><em>MS2 scans</em>: The number of MS2 (fragment) scans. Also shown as multiplier of MS1 scans.</li>
<li><em>Features</em>: The number of features captured in the featureXML file.</li>
<li><em>Feature intensity (Frac. TIC)</em>: The combined signal intensity attributed to a feature, as fraction of the total intensity of all peaks.</li>
<li><em>DB search matches</em>: The number of DB search matches (PSM's). Also as fraction of MS2 (success rate).</li>
<li><em>De novo identifications</em>: The number of <em>de novo</em> peptides. Also as fraction of MS2 (success rate).</li>
</ul>
<p><img alt="experiment_quality_metrics.PNG" src="../images/experiment_quality_metrics.png" /></p>
<h3 id="total-ion-current">Total Ion Current</h3>
<p>Show a profile of the total ion current (TIC) (and other parameters) over the retention time (RT). There are several options to change the data to display, or to combine spectral data with DB search matching <em>de novo</em> peptide identification data:</p>
<ul>
<li><strong>MS level</strong>: Profile either MS1 TIC or MS2 TIC over retention time.</li>
<li><strong>SMA window</strong>: Smoothening of the line plot is done by applying the Simple Moving Average (SMA) to the scan TIC values. The <em>SMA window</em> is the number of raw MS<em>x</em> scans to take for a single SMA value.</li>
<li><strong>Data reduction factor</strong>: In case of large scan numbers, the <em>data reduction factor</em> can be set to reduce the number of data points by taking each <em>n'th</em> data point while omitting the other points. This loses information but speeds up operation of the figure.</li>
<li><strong>Right axis param</strong>: Overlay the TIC/RT line with a second parameter using the right y-axis:<ul>
<li><em>DB search counts</em>: Plot the number of DB search matches as histogram.</li>
<li><em>De novo counts</em>: Plot the number of <em>de novo</em> identifications as histogram.</li>
<li><em>Peak count</em>: Plot a simple moving average of the number of peaks per MS1/MS2 scan.</li>
<li><em>Peak width (FWHM)</em>: Plot a simple moving average of elution times (at <em>Full Width Half Maximum</em>) of MS1 features.</li>
<li><em>Feature quality</em>: Plot a simple moving average of the calculated feature quality.</li>
<li><em>Ion injection time</em>: Plot a SMA of the MS1/MS2 <em>ion injection time</em>.</li>
<li><em>topN MS2</em>: Plot a SMA of the topN MS2 scans: The number of MS2 scans followed after a MS1 scan.</li>
</ul>
</li>
<li><strong>Confidence cutoff</strong>: (DB search counts <em>de novo</em> counts only) Only display DB search <em>de novo</em> peptide counts above the specified confidence threshold (only present if DB search <em>de novo</em> confidence selected as secondary parameter).</li>
<li><strong>int. cutoff</strong>: (Peak count only) Only consider peaks above a intensity threshold for visualization of the peak counts (see <em>secondary param</em>).</li>
</ul>
<p><img alt="QA_TIC_over_RT.PNG" src="../images/QA_TIC_over_RT.PNG" /></p>
<h3 id="mz-over-rt">m/z over RT</h3>
<p>Plot MS2 scans by their precursor m/z value and the retention time. This allows for a detailed profiling of the elution profile and can be used to evaluate the used LC gradient. Scans may be grouped by successful DB search <em>de novo</em> identification and filtered by a specified MS2 intensity cutoff.</p>
<p><img alt="QA_mz_over_RT.PNG" src="../images/QA_mz_over_RT.PNG" /></p>
<h3 id="ms2-ion-transmission">MS2 Ion Transmission</h3>
<p>Display the ion transmission efficiency for each MS2 scan in the spectral dataset. The MS2 Total Ion Current (TIC) (y-axis) is plotted against the precursor intensity (x-axis). Scans may be grouped by successful DB search <em>de novo</em> identification and a m/z cutoff may be applied to filter low m/z noise from the MS2 TIC.</p>
<p><img alt="QA_ms2_ion_transm.PNG" src="../images/QA_ms2_ion_transm.PNG" /></p>
<h3 id="scan-intensities">Scan Intensities</h3>
<p>The Scan Intensities figure shows the distribution of scans over the Total Ion Current range (TIC). The number of MS<em>x</em> scans at different TIC bins are counted. For MS2, scans may be grouped by successful DB search matched scans, <em>de novo</em> only identified scans (no DB search match), and unidentified scans. In addition, by checking <em>normalize bars</em>, the fraction of each scan category (color) at each TIC bin is displayed as opposed to the counts. </p>
<p><img alt="QA_int_dist.PNG" src="../images/QA_int_dist.PNG" /></p>
<h3 id="charge-distribution">Charge distribution</h3>
<p>A potential measure for contamination inside the MS run is the distribution of feature charges. Tryptic peptides will typically have charge 2-4 in the MS1 spectra. Therefore, a large fraction of high charged features (&gt;8) or singly charged features may be attributed to contamination in the sample. In this plot, the distribution of charge numbers across all observed features is displayed. These features may be partitioned in DB search matched, <em>de novo</em> only identified or unidentified fractions, if DB search and/or <em>de novo</em> data is included. This allows visualization of the success rate of peptide identification at a given charge state:</p>
<p><img alt="feature_charge_dist.PNG" src="../images/feature_charge_dist.png" /></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.92b07e13.min.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>