# Introduction

![[logo_with_name.PNG]]

MetaPepView is a metaproteomics dashboard interface designed for analysis of microbial communities and evaluation of metaproteomics experiments. It provides modules to visualize the taxonomic composition and expressed protein functions from microbial community data, as well as modules for evaluating metaproteomics experiment performance. It features a user-friendly, web-based interface that can be run locally on desktop PCs. MetaPepView enables users to perform large scale cross-sample quantitative analysis of community composition and protein functions, compare results from database searches and *de novo* annotations within samples, and evaluate the quality and coverage of metaproteomics experiments and reference databases. It is designed to be compatible with a variety of metaproteomics software tools.

Main features of MetaPepView include:

- Visualization of taxonomic compositions and expressed protein functions from metaproteomics experiments as obtained from various database search or *de novo* sequencing engines.
- DB independent taxonomy classification from *de novo* peptide sequences using the [Unipept](https://unipept.ugent.be/) API and comparison of the obtained *de novo* taxonomic composition with DB search results.
- Analysis of mass spectrometric performance from mzML raw files and evaluation of DB search and *de novo* matches from spectral data.
- Benchmarking of experimental performance with a reference dataset, consisting of a large number of reference experiments.

## Instrument support
The current version of MetaPepView is designed and has been tested for mass spectrometric raw data obtained from QE Orbitrap mass spectrometers, running in data-dependent acquisition (DDA) mode. Nevertheless, a broad support across mass spectrometry instruments is expected for taxonomic and functional analysis of communities from metaproteomics data. However, the spectral data analysis toolbox is not designed for data-independent acquisition (DIA) or PASEF based analysis strategies.

## PC requirements
- The dashboard is cross-platform and has been tested in Windows 10/11 and Linux.
- The dashboard can be installed with the python package manager `pip`, or setup with Docker (or another OCI compliant container manager) via the provided Docker image. See [[installation|Installation]] for more information.
- A minimum of 16 GB RAM is required for full operation of the dashboard, but 32 GB is recommended.
- The dashboard can be accessed through both Firefox (Gecko browser engine) and any Chromium based browser (Chrome, Edge, Brave, etc.).
- Internet connection is required to download taxonomy and function datasets, and for connection to the [Unipept](https://unipept.ugent.be/apidocs) API for *de novo* taxonomy classification.
- MetaPepView requires a recent Python version, make sure Python 3.11 or higher is installed

!!! note
	The browser engines support file import up to a certain threshold: (Chromium: ~300 MB, Firefox: ~1GB). Any larger file (which is common in mzML) will not be imported. **Trying to import a too large file into the dashboard will give no response from the browser, no warning will be shown.** To enable large file import, it is recommended to compress the files in *.zip* or *.tar.gz* format prior to import. If the mzML file is still too large, smaller mzML files could be generated by setting noise filters (top N signal threshold for example).
